Add script to run and compare CLI perf tests

dpodder opened the pull request:
`tests/Performance/run-perftests.py` is a Python3 script that fetches all
of the dependencies needed to run the perf tests in test/Performance
locally. It can run the perf tests for a single dotnet.exe, or run two
dotnet.exe instances sequentially and compare the results between them.

Basic usage for a single test:
```batch
  run-perftests.py --name "runid" ...\dotnet.exe
```

Usage for a comparison run:
```batch
  run-perftests.py --name "runid" ...\dotnet.exe --base ...\dotnet.exe
```

For more detailed usage:
```batch
    run-perftests.py -h
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dotnet/cli/2746)
<!-- Reviewable:end -->

dnfclas commented:
Hi __@dpodder__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

dpodder commented:
@brianrob PTAL

dnfclas commented:
__@dpodder__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

dpodder commented:
This is a follow-up to #2681

pakrym commented:
Why python? Cli build system is in C# we could just add another target and not add dependency on other language.

brianrob commented:
@pakrym, the goal is to have a lightweight script that can be run cross-platform.  Thus we chose Python, as this is what the .NET team has standardized on for cross-platform scripting.

pakrym commented:
CLI build script runs cross platform and is lightweight, just plain C#

dpodder commented:
This script doesn't just run the perf tests--it can, but it also runs perf tests for two builds and compares across them to compute speedup/slowdown. It's possible to reimagine this as a build target, but it'll have a dependency on an external (usually, previous) build, and so it made little sense to me to integrate it into the build system. Given that it's a standalone script, python made more sense.

lt72 commented on test/Performance/.gitignore:
```diff
@@ -0,0 +1,5 @@
+/ext/
+/*.html
+/*.csv
+/*.xml
```
excluding .xml files looks a bit aggressive, since there are .xml files in the repo... maybe better to trust that renaming *xml to *xml.bak will prevent git to pick those files up?

dpodder replied:
These are files that are generated during a perf test run. This particular .gitignore is intended to apply only to the subtree under test/Performance, but if *.xml is too aggressive even there then I'm happy to back it out. The only impact would be seeing untracked files after a run.

lt72 replied:
good then!

livarcocc commented:
How about using the dotnet CLI to run perf tests against the dotnet CLI? I totally agree with @pakrym. Don't really see a reason to take on this dependency.

brianrob commented:
I completely respect that there is a desire to build and test the CLI using the CLI.  From a performance perspective, we'd like to be part of that.

However, our tooling is not in a place for us to do that quickly, and the desire from both performance folks and those responsible for the CLI itself is to have something available fast that can be used by CLI developers to evaluate the performance of CLI changes.  In order to deliver on this, we choose to script the performance tests, and the xplat choice of scripting languages amongst the majority of .NET repos is Python - thus the choice.

Going forward, I belive that the right way to run performance tests for .NET core is to extend the functionality of the CLI, likely via the dotnet test functionality.  However, given that we've not even begun to investigate the work required to make this happen, it doesn't feel like the right way to unblock developers to test the performance of their local changes.

I hope this explains our rationale.

davidfowl commented on test/Performance/run-perftests.py:
```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/env python3
```
python???

brianrob commented:
Filed #2786 to track the longer-term story of running performance tests via dotnet test instead of this script.

dpodder commented:
Changes rebasing to latest & adding python2 support are ready. Will push just as soon as build completes (want to make sure everything is still clean.)

dpodder commented:
@brianrob @lt72 ported to work on python 2 (still works on python 3 as well). The framework targeting issue is still not resolved though, so please don't merge this yet.

brthor commented on test/Performance/run-perftests.py:
```diff
@@ -0,0 +1,238 @@
+        return 1
+
+if __name__ == "__main__":
+    sys.exit(main())
```
`main()` looks to return `None` on the happy path, will that produce 0 exit code here?

brthor replied:
Tried it out, and it does seem to produce a 0 exit code, may be good to be explicit though 😄

dpodder replied:
@brthor Easy enough, done. :smile:

brthor commented on test/Performance/run-perftests.py:
```diff
@@ -0,0 +1,238 @@
+    from_dir = kwargs['from_dir'] if 'from_dir' in kwargs else None
+    quiet = kwargs['quiet'] if 'quiet' in kwargs else False
+
+    quoted_args = map(lambda x: '"{x}"'.format(x=x) if ' ' in x else x, vargs)
```
quoting the args like this won't reflect the same escaping performed in the internals of subprocess.call, 

So the printed cmd_line may not repro the same issues, is that alright?

dpodder replied:
@brthor I agree that it would be better to match behavior. As it is, the cmd_line is only there for informational purposes, and so I struck a balance of matching the most common use case (spaces in args that are file paths), while keeping the code short.

If this is undesirable then I'm open to changing the output to not look like a command at all (e.g., just print the list of args directly, producing python-style output like `['arg1', 'arg2']`. I don't know of any way to get at subprocess's internal escaping implementation, and I really don't want to reimplement it here.

brthor replied:
Fair enough, this will cover 80% of the cases, and reimplenting that is a pain.. Sounds good.

brthor commented:
:shipit:

dpodder commented:
Thanks to @brianrob for help debugging the framework version issue! I've added the fix and rebased onto latest. All issues & PR feedback should now be addressed; ready for merge once CI is green and xunit-performance's [PR#126](https://github.com/Microsoft/xunit-performance/pull/126) is merged.

@lt72 fyi

brianrob commented:
Confirmed with @pakrym that this change works with #2820.  Ready to merge.
